common/debug/recover_stacktrace.go

```diff
@@ -0,0 +1,83 @@
+package debug
+
+import (
+	"fmt"
+	"os"
+	"path"
+	"path/filepath"
+	"runtime/debug"
+	"strings"
+	"syscall"
+	"time"
+
+	"github.com/ledgerwatch/erigon/log"
+)
+
+var sigc chan os.Signal
+
+func GetSigC(sig *chan os.Signal) {
+	sigc = *sig
+}
+
+func prettyTime() string {
+	time := fmt.Sprintf("%v", time.Now())
+	return strings.Replace(time[:19], " ", "-", 1)
+}
+
+func CheckForCrashes() {
+	ex, _ := os.Executable()
+	binPath := filepath.Dir(ex)
+	crashReportDir := path.Join(binPath[:len(binPath)-10], "crashreports")
+	f, err := os.Open(crashReportDir)
+	if err != nil {
+		log.Error(err.Error())
+		return
+	}
+	fileInfo, err := f.ReadDir(-1)
+	if err != nil {
+		log.Error(err.Error())
+		return
+	}
+	for _, v := range fileInfo {
+		if !v.IsDir() {
+			msg := fmt.Sprintf("Crashes From Previous Boots Detected. Find the stack trace in %v",
+				crashReportDir)
+			log.Warn(msg)
+			return
+		}
+	}
+}
+
+func RecoverStackTrace(err error, stopErigon bool, panicResult interface{}) error {
+	if panicResult != nil {
+		stack := string(debug.Stack())
+		switch typed := panicResult.(type) {
+		case error:
+			err = fmt.Errorf("%w, trace: %s", typed, stack)
+		default:
+			err = fmt.Errorf("%+v, trace: %s", typed, stack)
+		}
+		WriteStackTraceOnPanic(stack)
+		if stopErigon && sigc != nil {
+			sigc <- syscall.SIGINT
+		} else {
+			return err
```

Comment:
Why would we prefer continue after a panic than crush?

Reply:
There are functions like turbo/stages/stageloop.go: StageLoopSep where the panic is handled.

Reply:
We can introduce a decorator with a callback function to call on panic otherwise fail fast. In most cases it's better to fail then continue and corrupt data.